packages/devextreme/testing/testcafe/model/internal/widget.ts

```diff
@@ -1,18 +1,24 @@
-import { Selector, ClientFunction } from 'testcafe';
+import { ClientFunction } from 'testcafe';
+import { Selector } from '../../helpers/selector';
 import type { WidgetName } from '../../helpers/widgetTypings';
 import { isObject } from '../../../../js/core/utils/type';
 
 function getComponentInstance(
   selector: Selector,
   name?: string,
 ): () => Promise<unknown> {
+  const isShadowDom = process.env.shadowDom === 'true';
+
   return ClientFunction(
     () => {
       const $widgetElement = $(selector());
       const elementData = $widgetElement.data();
-      const widgetNames = elementData.dxComponents;
+      const widgetNames = elementData?.dxComponents ?? [];
 
       if (name) {
+        if (isShadowDom) {
+          return (window as any).DevExpress.ui[name].getInstance($widgetElement.get(0));
+        }
```

Comment:
Looks like logic inside `if (isShadowDom)` should work for non-shadow mode too. If so, let's leave only this logic
Btw, why `elementData` doesn't work? It's totally not working inside shadow dom?

Reply:
discussed, `else` doesn't work for all widgets (like private Scrollable), so we need to leave both